### PR TITLE
Add new type system functions in the frontend

### PR DESCRIPTION
**Pull request by thehabbos007**
## 🌟 What is the purpose of this PR?

This PR introduces a simple environment in which we can start making use of the new type system types.
The existing list of BP functions offered to blocks has been changes in a new wrapper `useBlockProtocolFunctionsWithOntology`, which adds new functions. The name is very much a WIP/"internal" name, and the operation list is meant to replace the existing BP functions when the BP package updates with the type system type declarations.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1200211978612931/1202813127213855/f) _(internal)_

## 🔍 What does this change?

-  Adds the following functions
    ```
    aggregateDataTypes
    getDataType
    createPropertyType
    aggregatePropertyTypes
    getPropertyType
    updatePropertyType
    createEntityType
    aggregateEntityTypes
    getEntityType
    updateEntityType
    createLinkType
    aggregateLinkTypes
    getLinkType
    updateLinkType
    ```
    to the list of Block Protocol functions we can provide to blocks
- Adds a basic page under `/type-editor` to provide a place to start using the type

## 📜 Does this require a change to the docs?

- The docs for x need updating to explain that y

## ⚠️ Known issues

-  We are currently not adding the user as a react context and using it in the frontend, I decided to not spend time figuring this out in this piece of work.
-  The GQL metadata is leaked to the BP response, but the fields are untyped in TS. there should be a mapping function between GQL and the block

## 🐾 Next steps

- Integrate with the Type System package residing in the BP repo
- Start developing against the new set of functions (type editors)

## 🛡 What tests cover this?

- None

## ❓ How to test this?
1.  Checkout the branch / view the deployment
1.  Run external services: `yarn external-services up`
1.  Run the frontend/backend: `yarn dev` from the root
1.  Login at `http://localhost:3000/` with a dev user (e.g. `alice@example.com:password`)
1.  Go to `http://localhost:3000/type-editor`

## 📹 Demo

image

### CODE REVIEW

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateEntityTypes.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+            {
+              code: "INVALID_INPUT",
+              message: "'data' must be provided for aggregateEntityTypes",
+            },
+          ],
+        };
+      }
+
+      const response = await aggregateFn({
+        query: getAllLatestEntityTypesQuery,
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateEntityTypes.ts:33]: Do we want a @todo here?

File: packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
+import { EmbedderGraphMessageCallbacks } from "@blockprotocol/graph";
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts:1]: Should we add a top-level doc on this one as well expressing its purpose and when we're going to delete it

File: packages/hash/frontend/src/pages/type-editor/index.page.tsx [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    if (!kratosSession && !user) {
+      void router.push("/login");
+    }
+  }, [loading, router, user, kratosSession]);
+
+  /** @todo use the User's accountId here. */
+  const functions = useBlockProtocolFunctionsWithOntology("..");
+
+  const getType = useCallback(
```
Comment (Alfred-Mountfield) [inline packages/hash/frontend/src/pages/type-editor/index.page.tsx:33]: Can you structure this file a bit more clearly to show what's part of the template/boilerplate and what's there for demonstration purposes? (Even just a comment delineating the sections or something)

**Review by Alfred-Mountfield — COMMENTED** [review]
Generally looks good to me. Obviously a lot of code that is there for temporary purposes so I'm not going to be super nitpicky with it. I think we should only merge once we've spoken to Nate about the index file and what dev will look like

File: packages/hash/frontend/codegen.yml [hunk 4]
```diff
@@ -11,9 +11,15 @@ generates:
     config:
       skipTypename: true
       scalars:
+        # TODO Switch out this direct dependency on the HASH Graph SDK with
+        #   the a shared type package
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/frontend/codegen.yml]: Addressed in fd02a90db990e1cca9d127ffe7afd0d7698afe52

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/ontology-types-shim.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
+import { MessageCallback } from "@blockprotocol/core";
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/ontology-types-shim.ts:9]: Addressed in 4ceac40dbeb308030fdf201cd530544bfbbab32a

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/ontology-types-shim.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  updateEntityType: UpdateEntityTypeMessageCallback;
+  createLinkType: CreateLinkTypeMessageCallback;
+  aggregateLinkTypes: AggregateLinkTypesMessageCallback;
+  getLinkType: GetLinkTypeMessageCallback;
+  updateLinkType: UpdateLinkTypeMessageCallback;
+};
+
+/* Shared types */
+
+type Response<N extends string, T> = {
```
[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/ontology-types-shim.ts:49]: Addressed in e1c72404b403415d9e4ce67bd6b5dd01f162e1f1

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateEntityTypes.ts (hunk 1 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateEntityTypes.ts:34]: Addressed in 362495ae96a6af4cd56c0a07b6223b64399e702a

File: packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts (hunk 2 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/pages/type-editor/blockprotocol-ontology-functions-hook.ts:9]: Addressed in 68e724d050c21833ff217ae3c6bb53874abeed5b

File: packages/hash/frontend/src/pages/type-editor/index.page.tsx (hunk 3 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/pages/type-editor/index.page.tsx:29]: Addressed in 2a01ba68158085cac0e00a2705d8f1f105c21ed4
Added a button to create a property type as well in the demo page

Comment (Alfred-Mountfield) [reply packages/hash/frontend/src/pages/type-editor/index.page.tsx:29]: Better, thank you!

**Review by Alfred-Mountfield — DISMISSED** [review]
Looks good to me, there's a lot of code so I'm fairly certain some bugs are sneaking past us but as this is mostly to enable a dev environment I'm quite happy for us to tackle them as we find them rather than worrying about perfection here.

Probably still worth a _quick_ sanity check from @CiaranMn on the shimmed methods and such

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateDataTypes.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      return {
+        data: {
+          results: response.data.getAllLatestDataTypes,
+        },
+      };
+    },
+    [aggregateFn],
+  );
+
+  return { aggregateDataTypes };
```
Comment (nathggns) [inline packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateDataTypes.ts:59]: Curious why we don't just return the function itself?

**Review by nathggns — APPROVED** [review]

File: packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateDataTypes.ts (hunk 7 above)
Comment (thehabbos007) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateDataTypes.ts:59]: This was more or less convention from the other functions. I assume it helps enforce that we don't rename the methods when using the "hook"

Comment (nathggns) [reply packages/hash/frontend/src/components/hooks/blockProtocolFunctions/ontology/useBlockProtocolAggregateDataTypes.ts:59]: The convention with hooks is to let the consumer name the return value. There are *some* exceptions. I don't think we need to change this now but I do think we should look into it across the board.

**Comment by thehabbos007** [comment]
Added asana links to todos where relevant

**Review by nathggns — APPROVED** [review]

**Pull request merged**